src/ExactDiagonalization/basis_bfs.jl

```diff
@@ -0,0 +1,398 @@
+
+    # addresses already visited and addresses visited in the last layer
+    seen = Set(basis)
+    last_seen = empty(seen)
+
+    # addresses to visit and their index
+    curr_frontier = collect(zip(basis, eachindex(basis)))
+    next_frontier = empty(curr_frontier)
+
+    # map from address to index
+    mapping = Dict(zip(basis, eachindex(basis)))
+
+    # builders store task-local storage and choose whether a matrix is to be built or not
+    builders = [Builder(; col_hint) for _ in 1:max_tasks]
+    result_accumulator = init_accumulator(Builder)
+
+    depth = 0
+    early_stop = false
+    while !isempty(curr_frontier)
+        depth += 1
+        early_stop = length(basis) > stop_after || depth > max_depth
+
+        # We can stop here when not constructing the matrix. If we are, we still need
+        # to do another round to evaluate the columns.
+        if Builder <: BasisBuilder && early_stop
+            break
+        end
+
+        # Split the workload into chunks and spawn a task for each. These now run
+        # asynchronously while the main thread continues.
+        split_frontier = UniformSplit(curr_frontier, min_batch_size, max_tasks)
+        tasks = map(enumerate(split_frontier)) do (i, sub_frontier)
+            Threads.@spawn builders[$i]($operator, $sub_frontier, seen)
+        end
+
+        # Procesess the tasks in order they were spawned. This is fine as we've used more
+        # tasks than threads.
+        for (task, builder) in zip(tasks, builders)
+            wait(task)
+
+            result = builder.frontier
+            for k in result
+                if k ∉ last_seen
+                    push!(last_seen, k)
+                    if !early_stop && (isnothing(filter) || filter(k))
+                        push!(basis, k)
+                        push!(next_frontier, (k, length(basis)))
+                        mapping[k] = length(basis)
+                    end
+                end
+            end
+
+            # update matrix if needed
+            update_accumulator!(result_accumulator, builder, mapping)
+        end
+
+        # This had to wait until now so we don't get into data races
+        union!(seen, last_seen)
+        curr_frontier, next_frontier = next_frontier, curr_frontier
+        empty!(next_frontier)
+        empty!(last_seen)
+    end
+
+    return finalize_accumulator!(result_accumulator, basis, sort; sort_kwargs...)
+end
+
+function _address_to_basis(operator, addr_or_basis)
+    if addr_or_basis isa AbstractVector || addr_or_basis isa Tuple
+        check_address_type(operator, eltype(addr_or_basis))
+        return collect(addr_or_basis)
+    else
+        check_address_type(operator, typeof(addr_or_basis))
+        return [addr_or_basis]
+    end
+end
+
+"""
+    build_basis(
+        ham, address=starting_address(ham);
+        cutoff, filter, sizelim, sort=false, kwargs...
+    ) -> basis
+    build_basis(ham, addresses::AbstractVector; kwargs...)
```

Comment:
It would be nice to support a `minimum_size` keyword (or something equivalent) as just introduced in the Excited States PR.